vmm/src/landlock.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright © 2024 Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::convert::TryFrom;
+use thiserror::Error;
+
+use crate::vm_config::LandlockConfig;
+use landlock::{
+    path_beneath_rules, Access, AccessFs, BitFlags, Ruleset, RulesetAttr, RulesetCreated,
+    RulesetCreatedAttr, RulesetError, ABI,
+};
+use std::io::Error as IoError;
+use std::path::PathBuf;
+
+#[derive(Debug, Error)]
+pub enum LandlockError {
+    /// All RulesetErrors from Landlock library are wrapped in this error
+    #[error("Error creating/adding/restricting ruleset: {0}")]
+    ManageRuleset(#[source] RulesetError),
+
+    /// Error opening path
+    #[error("Error opening path: {0}")]
+    OpenPath(#[source] IoError),
+
+    /// Invalid Landlock access
+    #[error("Invalid Landlock access: {0}")]
+    InvalidLandlockAccess(String),
+
+    /// Invalid Path
+    #[error("Invalid path")]
+    InvalidPath,
+}
+
+// https://docs.rs/landlock/latest/landlock/enum.ABI.html for more info on ABI
+static ABI: ABI = ABI::V3;
+
+pub struct LandlockAccess {
+    access: BitFlags<AccessFs>,
+}
+
+impl TryFrom<&str> for LandlockAccess {
+    type Error = LandlockError;
+
+    fn try_from(s: &str) -> Result<LandlockAccess, LandlockError> {
+        let mut access = BitFlags::<AccessFs>::empty();
+        for c in s.chars() {
+            match c {
```

Comment:
Please note that using `from_*(abi)` means that future version of Cloud Hypervisor may be required to stick to the same Landlock ABI to avoid unexpected denied access. I suggest to document that 'r', 'w', and 'x' will translate to more restrictions in the future (e.g. I think we'll want 'w' to deny xattr changes, which is not implemented yet), and users need to be aware of this semantic. For instance, it would make sense to document that these access right will be similar to unix permissions eventually.

Reply:
I will add a blurb to this effect in landlock.md

Reply:
Below is the blurb added to landlock.md:

```
Under the scope of read and write access Landlock could allow some additional accesses (eg: write access allows xattr changes). Eventually, Landlock will only allow accesses similar to unix permissions.
```
Please check if this needs to be updated.